2841: Resolve CollectionUtils

kduske opened the pull request:
This is part of #2844

kduske commented:
Build time summary:

Before:
```
real	1m42.984s
user	10m41.875s
sys	1m32.771s
```

After:
```
real	1m40.919s
user	10m33.685s
sys	1m32.479s
```

ericwa commented on common/src/Model/CollectMatchingNodesVisitor.h:
```diff
@@ -101,14 +102,16 @@ namespace TrenchBroom {
                 ++cur;
             }
-            return result;
+            return result.release_data();
```
what's this?

kduske replied:
It returns the underlying vector of the vector_set by moving, which is safe here since its not reused.

ericwa commented on common/src/IO/ObjSerializer.h:
```diff
@@ -53,10 +52,13 @@ namespace TrenchBroom {
-                    typename Map::iterator indexIt = MapUtils::findOrInsert(m_map, v, m_list.size());
-                    const size_t index = indexIt->second;
-                    if (index == m_list.size())
+                    const auto [it, success] = m_map.try_emplace(v, m_list.size());
```
.insert(

kduske replied:
Yes, here it makes sense.

ericwa commented on common/src/IO/LegacyModelDefinitionParser.cpp:
```diff
@@ -92,25 +91,20 @@ namespace TrenchBroom {
-                do {
-                    expect(status, MdlToken::String | MdlToken::Word, token = m_tokenizer.peekToken());
-                    if (token.hasType(MdlToken::String))
-                        modelExpressions.push_back(parseStaticModelDefinition(status));
```
necessary?

kduske replied:
see above

ericwa commented on common/src/Renderer/ShaderManager.cpp:
```diff
@@ -19,58 +19,57 @@
-            ShaderProgram* program = createProgram(config);
-            m_programs.insert(ShaderProgramCacheEntry(&config, program));
-            return *program;
+            auto result = m_programs.emplace(&config, createProgram(config));
```
this is weird that is stores a pointer to `config`, given that config is passed by const reference. I know it was like that before, I'd add a fixme to change the function signature to `program(const ShaderConfig* config)`

kduske replied:
Agreed, I'm changing it right away though.

kduske replied:
on second thought, it makes sense if you look at how we are using it; the shader configs are declared as constants in Shaders.h. It's a bit iffy, but if we take a pointer, we have to pass the address everywhere, which makes using the API odd. The fact that it's stored by pointer in the map seems like an implementation detail to me.

ericwa commented on lib/kdl/include/kdl/vector_utils.h:
```diff
@@ -0,0 +1,381 @@
+#include <type_traits> // for std::less
+#include <vector>
+
+namespace kdl {
```
I commented before but I'd suggest a `vec_` or `vec::` prefix on all of these.

kduske replied:
I agree, I'll go with `vec_` and `map_`.

ericwa commented on lib/kdl/include/kdl/set_adapter.h:
```diff
@@ -0,0 +1,925 @@
+         */
+        void clear() {
+            m_data.clear();
+            assert(check_invariant());
```
check_invariant() should be behind a #ifdef SET_ADAPTER_DEBUG, IMO. 
Having all of the methods O(n) times slower in debug builds is annoying, especially if we're going to be using this a lot.

kduske replied:
done

ericwa commented on lib/kdl/include/kdl/map_utils.h:
```diff
@@ -0,0 +1,219 @@
+     * @return an int indicating the result of the comparison
+     */
+    template<typename K, typename V, typename C, typename D = std::less<V>>
+    int lexicographical_compare(const std::map<K, V, C>& map1, const std::map<K, V, C>& map2, const D& value_cmp = D()) {
```
map_lexicographical_compare?

kduske replied:
done

ericwa commented on lib/kdl/include/kdl/map_utils.h:
```diff
@@ -0,0 +1,219 @@
+    template<typename K, typename V, typename C, typename D = std::less<V>>
+    int lexicographical_compare(const std::map<K, V, C>& map1, const std::map<K, V, C>& map2, const D& value_cmp = D()) {
+        const auto key_cmp = C();
+        return kdl::lexicographical_compare(std::begin(map1), std::end(map1), std::begin(map2),
```
kdl::range_lexicographical_compare? kdl::iterator_lexicographical_compare?

kduske replied:
done

ericwa commented on lib/kdl/include/kdl/map_utils.h:
```diff
@@ -0,0 +1,219 @@
+#include <map>
+#include <vector>
+
+namespace kdl {
```
same comment as for vectors, these should either all start with map_ or be in `kdl::map::`.

kduske replied:
done

ericwa commented on lib/kdl/include/kdl/vector_utils.h:
```diff
@@ -0,0 +1,381 @@
+        kdl::delete_all(v, deleter);
+        v.clear();
+    }
+}
```
I'd suggest adding some signed-int accessors:
`T at(const std::vector<T>& v, int index);`
`T at(const std::vector<T>& v, int64_t index);`

also generic collection size getters:
`int size_int(const C& collection);`
`int64_t size_int64(const C& collection);`

Whenever we have static_casts from signed ints <-> size_t, this is a code smell, and there's a lot of these in TB!
(this should be a separate issue I guess.)

kduske replied:
I'll add these methods, but I think most of the type conversions are because Qt uses int for indexing / size.

kduske replied:
Which I believe is the actual code smell here ;-)

kduske replied:
done

ericwa replied:
cool, thanks. 

I blame the stl using unsigned, though. The static casts to work around this crop up in TB not just in code interacting with Qt.

there's a video that's worth checking out where this topic comes up a few times, that is the best discussion of this I've seen, links in this post 
https://github.com/ericniebler/stl2/issues/182#issuecomment-287683189

kduske replied:
It looks like everyone is disagreeing with me ;-)

ericwa commented on common/src/View/VertexToolBase.h:
```diff
@@ -323,7 +325,7 @@ namespace TrenchBroom {
             template <typename HH>
             void renderHandles(const std::vector<HH>& handles, Renderer::RenderService& renderService, const Color& color) const {
                 renderService.setForegroundColor(color);
-                renderService.renderHandles(VectorUtils::cast<typename HH::float_type>(handles));
```
kdl::vec_elements_cast?

kduske replied:
done

ericwa commented on common/src/View/FileTextureCollectionEditor.cpp:
```diff
@@ -174,7 +175,7 @@ namespace TrenchBroom {
             auto collections = document->enabledTextureCollections();
 
             const auto index = static_cast<size_t>(m_collections->currentRow());
-            VectorUtils::swapPred(collections, index);
```
needs an ensure(index > 0, "???");

kduske replied:
I have opted to add an if here. I used to think it's better to catch the errors and fix the cause, but in some cases it's just better to be defensive IMO.

ericwa commented on common/src/View/DuplicateNodesCommand.cpp:
```diff
@@ -53,15 +55,17 @@ namespace TrenchBroom {
 
                     Model::Node* parent = original->parent();
                     if (cloneParent(parent)) {
-                        auto insertPos = MapUtils::findInsertPos(newParentMap, parent);
```
I still don't understand what is going on here, why is it using lower_bound and what does std::prev(hint) represent?

kduske replied:
The code looks like this because I wanted to avoid doing two lookups in the map, one for checking if the parent is already cloned and then again on insertion if it turned out that it hasn't been cloned yet. But you are right in that it makes the code more complicated at little gain. I'll change it so that its easier to understand.

kduske replied:
done

kduske commented on common/src/Assets/AttributeDefinition.cpp:
```diff
@@ -243,7 +244,12 @@ namespace TrenchBroom {
+                return nullptr;
+            } else {
+                return &(*it);
+            }
```
done

kduske commented on common/src/Assets/EntityDefinition.cpp:
```diff
@@ -89,7 +88,12 @@ namespace TrenchBroom {
 
         const FlagsAttributeDefinition* EntityDefinition::spawnflags() const {
-            return static_cast<FlagsAttributeDefinition*>(VectorUtils::findIf(m_attributeDefinitions, FindSpawnflagsDefinition()).get());
+            auto it = std::find_if(std::begin(m_attributeDefinitions), std::end(m_attributeDefinitions), FindSpawnflagsDefinition());
```
done

kduske commented on common/src/Assets/ModelDefinition.cpp:
```diff
@@ -91,12 +91,12 @@ namespace TrenchBroom {
             EL::ExpressionBase::List cases;
-            cases.push_back(m_expression.clone());
-            cases.push_back(other.m_expression.clone());
+            cases.emplace_back(m_expression.clone());
```
In this case, `clone()` returns a raw pointer, but the list contains std::unique_ptr, so in this case we need to call `emplace` as push_back would give a type error. The alternative would be `push_back(std::unique_ptr<ExpressionBase>(m_expression->clone()))` which is a bit too wordy for my taste here.

ericwa replied:
Ahh that makes a lot more sense now. I think the problem is I'm reading on github, and with the combination of emplace_back and the type alias EL::ExpressionBase::List, I didn't realize std::unique_ptr was involved

ericwa replied:
whatever you think is best. would `push_back(std::make_unique(m_expression->clone()))` work/be a good compromise?

kduske replied:
I don't think make_unique works that way. You'd have to pass it the arguments of the type that will be wrapped by the unique_ptr, which in this case would be unknown.

kduske replied:
I agree with the abseil tip: If both work with the same arguments, use push_back / insert. But here that's not the case.

kduske commented on common/src/EL/Expression.cpp:
```diff
@@ -164,44 +167,42 @@ namespace TrenchBroom {
-            for (const ExpressionBase* element : m_elements)
-                clones.push_back(element->clone());
+            for (const auto& element : m_elements) {
+                clones.emplace_back(element->clone());
```
Same argument as above.

kduske commented on common/src/EL/Value.cpp:
```diff
@@ -1029,7 +1032,7 @@ namespace TrenchBroom {
                     switch (rhs.type()) {
                         case ValueType::Array:
-                            return Value(VectorUtils::concatenate(lhs.arrayValue(), rhs.arrayValue()));
+                            return Value(kdl::concat(lhs.arrayValue(), rhs.arrayValue()));
```
I'll introduce a namespace `vec` for the vector_utils methods, likewise for map_utils.

kduske replied:
maybe a prefix is better, because otherwise we end up with a function named `map::union` which is a keyword ;-)

kduske commented on common/src/IO/ELParser.cpp:
```diff
@@ -313,15 +312,15 @@ namespace TrenchBroom {
             if (!m_tokenizer.peekToken().hasType(ELToken::CBracket)) {
                 do {
-                    elements.push_back(parseExpressionOrAnyRange());
+                    elements.emplace_back(parseExpressionOrAnyRange());
```
Same situation as before, the function returns a raw pointer.

kduske commented on common/src/IO/ELParser.cpp:
```diff
@@ -460,13 +457,13 @@ namespace TrenchBroom {
             if (token.hasType(ELToken::SimpleTerm)) {
                 do {
-                    subExpressions.push_back(parseExpression());
+                    subExpressions.emplace_back(parseExpression());
```
see above

kduske commented on common/src/IO/ELParser.cpp:
```diff
@@ -367,13 +366,13 @@ namespace TrenchBroom {
             if (!m_tokenizer.peekToken().hasType(ELToken::CBracket)) {
                 do {
-                    elements.push_back(parseExpressionOrRange());
+                    elements.emplace_back(parseExpressionOrRange());
```
see above

kduske commented on common/src/IO/FileSystem.cpp:
```diff
@@ -164,10 +165,10 @@ namespace TrenchBroom {
             }
 
-            VectorUtils::sortAndRemoveDuplicates(result);
+            kdl::sort_and_make_unique(result);
```
yes, agree

kduske replied:
done

kduske commented on common/src/Renderer/TexturedIndexRangeMap.cpp:
```diff
@@ -43,7 +42,8 @@ namespace TrenchBroom {
         IndexRangeMap::Size& TexturedIndexRangeMap::Size::findCurrent(const Texture* texture) {
             if (!isCurrent(texture)) {
-                m_current = MapUtils::findOrInsert(m_sizes, texture, IndexRangeMap::Size());
+                const auto result = m_sizes.try_emplace(texture);
```
It's the same situation as before though:

```
                const auto result = m_sizes.insert(texture);
```

will not compile, we'd have to create an unnecessary instance of the value type. With try_emplace, we can get away with just passing the key and having the value created if and only if the map doesn't yet have an entry for the key.

kduske commented on common/src/Renderer/TexturedIndexArrayMap.cpp:
```diff
@@ -46,7 +45,8 @@ namespace TrenchBroom {
         IndexArrayMap::Size& TexturedIndexArrayMap::Size::findCurrent(const Texture* texture) {
             if (!isCurrent(texture)) {
-                m_current = MapUtils::findOrInsert(m_sizes, texture, IndexArrayMap::Size());
+                const auto result = m_sizes.try_emplace(texture);
```
see above

kduske commented on common/src/Renderer/IndexRangeMap.cpp:
```diff
@@ -160,7 +160,8 @@ namespace TrenchBroom {
             auto it = m_data->end();
             if (m_dynamicGrowth) {
-                it = MapUtils::findOrInsert(*m_data, primType);
+                auto result = m_data->try_emplace(primType);
```
see above

kduske commented on common/src/Model/EntityAttributes.h:
```diff
@@ -139,7 +139,7 @@ namespace TrenchBroom {
             EntityAttribute::List listFromQueryResult(const std::vector<IndexValue>& matches) const;
         public:
-            const std::set<AttributeName> names() const;
+            const std::vector<AttributeName> names() const;
```
I will double check this.

kduske replied:
The constraint could have been violated by `setAttributes`, I have added a check to ensure that duplicate attributes are not added via that method.

kduske commented on common/src/Renderer/TextureFont.cpp:
```diff
@@ -145,7 +145,7 @@ namespace TrenchBroom {
             void makeQuads(const String& str, const float x) {
                 const auto offset = m_offset + vm::vec2f(x, m_y);
-                VectorUtils::append(m_vertices, m_font.quads(str, m_clockwise, offset));
+                kdl::append(m_vertices, m_font.quads(str, m_clockwise, offset));
```
done

kduske commented on common/src/Model/TagMatcher.cpp:
```diff
@@ -93,7 +95,7 @@ namespace TrenchBroom {
                 texture = matchingTextures.front();
             } else {
-                const auto options = VectorUtils::map(matchingTextures, [](const auto* current) { return current->name(); });
+                const auto options = kdl::transform(matchingTextures, [](const auto* current) { return current->name(); });
```
done

kduske commented:
@ericwa thank you for the in-depth review! I have addressed your comments and will merge once the builds turn green. If there is anything that I haven't resolved in the way you intended, we can fix it in a separate issue.
